### PR TITLE
Workarounds for limit + order by queries 

### PR DESCRIPTION
**Pull request by kexinrong**

### CODE REVIEW

File: k8s-scheduler/src/main/java/com/vmware/dcm/AutoScope.java [hunk 1]
```diff
@@ -19,18 +21,134 @@
...
+                    records.add(r);
+                } else {
+                    final int index = helper.addToTopk(records, r);
+                    if (index >= 0) {
+                        records.set(index, r);
+                    }
+                }
+                topk.put(group, records);
+            } else {
+                final List<Record> records = topk.getOrDefault(group, new ArrayList<>());
```
Comment (lalithsuresh) [inline k8s-scheduler/src/main/java/com/vmware/dcm/AutoScope.java:111]: The logic needs to replace existing elements from the full list when deletions happen.

**Review by lalithsuresh — CHANGES_REQUESTED** [review]
Would be good to also add a few tests to make sure this works correctly. In particular, tests to add, remove records, and test the boundary cases of a view having size equal to, less than and greater than K.

File: k8s-scheduler/src/main/java/com/vmware/dcm/Scheduler.java [hunk 2]
```diff
@@ -457,22 +466,35 @@ private void requeue(final Result<? extends Record> unassignedPods) {
...
-                        LOG.info(String.format("Scoping Optimization: Replace %s with %s", t.getName(), augView));
                         toFetch = table(augView);
                     } else {
                         toFetch = t;
                     }
                     if (dbConnectionPool instanceof DDlogDBConnectionPool) {
-                        return ((DDlogDBConnectionPool) dbConnectionPool).getProvider()
-                                .fetchTable(toFetch.getName());
+                        final DDlogJooqProvider provider = ((DDlogDBConnectionPool) dbConnectionPool).getProvider();
+                        final Result<Record> augResult = provider.fetchTable(toFetch.getName());
```
Comment (lalithsuresh) [inline k8s-scheduler/src/main/java/com/vmware/dcm/Scheduler.java:483]: I might be missing something. Wouldn't `augResult` in line 483 already have all the rows in it (that is, the query without the topK filter)? And if so, wouldn't the condition in line 488 always be true?

File: k8s-scheduler/src/test/java/com/vmware/dcm/SchedulerTest.java [hunk 3]
```diff
@@ -1245,6 +1245,87 @@ public void testSchedulerDebugDump() {
...
+        final Iterator<Integer> memCapIt = nodeMemoryCapacities.iterator();
+        final Iterator<Integer> storageCapIt = nodeStorageCapacities.iterator();
+        final Iterator<Integer> podsCapIt = nodepodsCapacities.iterator();
+        final HashMap<String, List<Integer>> expected = new HashMap<>();
+        expected.put("cpu", (List<Integer>) args.get(5));
+        expected.put("memory", (List<Integer>) args.get(6));
+        expected.put("ephemeral-storage", (List<Integer>) args.get(7));
+        expected.put("pods", (List<Integer>) args.get(8));
+
+        final int limit = 3;
```
Comment (lalithsuresh) [inline k8s-scheduler/src/test/java/com/vmware/dcm/SchedulerTest.java:1270]: Not sure, but is this limit large enough for the test (that is, does it only pull a subset of nodes out)?

**Review by lalithsuresh — CHANGES_REQUESTED** [review]

File: k8s-scheduler/src/test/java/com/vmware/dcm/SchedulerTest.java (hunk 3 above)
Comment (kexinrong) [reply k8s-scheduler/src/test/java/com/vmware/dcm/SchedulerTest.java:1270]: Yep it's large enough. The test case has 10 nodes.

File: k8s-scheduler/src/main/java/com/vmware/dcm/Scheduler.java (hunk 2 above)
Comment (kexinrong) [reply k8s-scheduler/src/main/java/com/vmware/dcm/Scheduler.java:483]: Before this PR, in AutoScope, we create spare_capacity_per_node_augment = (domain restricting queries) UNION (top K queries). Right now in AutoScope, the augment view only contains the domain restricting conditions and we are performing the union in the for loop on line 487. Therefore line 488 is not always true.

Comment (lalithsuresh) [reply k8s-scheduler/src/main/java/com/vmware/dcm/Scheduler.java:483]: Ah, got it. Thanks.

In my head I was thinking of the solution where we were applying topK to multiple domain restricting views, and union-ing the results. This is not that case, but is instead, as you say, domain restricting views UNION topK.

**Review by lalithsuresh — APPROVED** [review]
